patches/api/0387-Add-damage-source-wrapper.patch

```diff
@@ -0,0 +1,286 @@
++     * Also the target receive no knockback
++     */
++    EXPLOSION,
++    /**
++     * Apply a fire damage, the target will burn and take
++     * fire damage if the game-rule {@link org.bukkit.GameRule#FIRE_DAMAGE} is enabled
++     * and that the player hasn't drunk a fire resistance potion.
++     * An armor with the fire protection enchantment can reduce the
++     * damage made.
++     *
++     * Some entities can have a special reaction to this damage:
++     * - {@link org.bukkit.entity.Item} can die immediately
++     * - {@link org.bukkit.entity.minecart.ExplosiveMinecart} can explode immediately
++     *
++     * Some entities (from nether mostly) can have developed an immunity to the fire
++     */
++    FIRE,
++    MAGIC,
++    FALL,
++    GENERIC
++}
+diff --git a/src/main/java/com/destroystokyo/paper/entity/damage/DamageBypass.java b/src/main/java/com/destroystokyo/paper/entity/damage/DamageBypass.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..f86b96deaa3940494e9366457d3203b91be83e04
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/entity/damage/DamageBypass.java
+@@ -0,0 +1,17 @@
++package com.destroystokyo.paper.entity.damage;
++
++public enum DamageBypass {
++    ARMOR(true), MAGIC(true), INVULNERABILITY, ENCHANTMENT; // Order is important for bit op
++
++    public final int bitmask;
++    public final boolean changeExhaustion;
++
++    DamageBypass() {
++        this(false);
++    }
++
++    DamageBypass(boolean changeExhaustion) {
++        bitmask = (1 << ordinal());
++        this.changeExhaustion = changeExhaustion;
++    }
++}
+diff --git a/src/main/java/com/destroystokyo/paper/entity/damage/DamageOrigin.java b/src/main/java/com/destroystokyo/paper/entity/damage/DamageOrigin.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..b41b1fd60bbb6362dd05ef661b3f54df92c167ac
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/entity/damage/DamageOrigin.java
+@@ -0,0 +1,47 @@
++package com.destroystokyo.paper.entity.damage;
++
++import net.kyori.adventure.text.Component;
++import org.bukkit.entity.*;
++import org.jetbrains.annotations.Nullable;
++
++public interface DamageOrigin {
++
++    boolean isSweep();
++
++    boolean isCritical();
++
++    boolean isFromProjectile();
++
++    boolean isFromExplosion();
++
++    boolean isFromFire();
++
++    boolean isFromMagic();
++
++    boolean isFromFalling();
++
++    boolean willNotAggro();
++
++    boolean willScalesWithDifficulty();
++
++    default boolean canBypass(DamageBypass... damageType) {
++        for(DamageBypass type : damageType) {
++            if(!canBypass(type)) return false;
++        }
++        return true;
++    }
++
++    boolean canBypass(DamageBypass damageType);
++
++    boolean canDamageHelmetArmor();
++
++    float getFoodExhaustion();
++
++    @Nullable
++    Entity getDirectEntity();
++
++    @Nullable
++    Entity getSource();
++
```

Comment:
is there anywhere where these entities are used in normal DamageSources? It might be useful to move these to EntityDamageOrigin/ IndirectEntityDamageOrigin .

Reply:
I don't think but this is convenient because the event return a DamageOrigin and not an EntityDamageOrigin (so plugin don't need to cast anything) maybe i can move this in the EntityDamageOrigin class and cast the DamageOrigin in the EntityDamageByEntityEvent ? But i'm not sure if bukkit really track well the difference between EntityDamageByEntityEvent and EntityDamageEvent in the factory class

Reply:
I mean it's mostly because yeah these will always be null in the case that it isn't an EntityDamageSource or IndirectDamageSource. 
But, your call what you would like to do with this.